version/src/test/java/io/smallrye/common/version/VersionRangeTest.java

```diff
@@ -0,0 +1,98 @@
+        assertFalse(versionRange.test("2.0.0"));
+        assertFalse(versionRange.test("2.0.1"));
+    }
+
+    @Test
+    void testVersionRangeWithLowerBoundExclusive() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "(1.0,2.0]");
+        assertFalse(versionRange.test("1.0.0"));
+        assertTrue(versionRange.test("1.1.0"));
+        assertTrue(versionRange.test("1.899.0"));
+        assertTrue(versionRange.test("2.0"));
+        assertTrue(versionRange.test("2.0.0"));
+        assertFalse(versionRange.test("2.0.1"));
+    }
+
+    @Test
+    void testVersionRangeWithUpperBoundExclusive() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "[1.0,2.0)");
+        assertTrue(versionRange.test("1.0.0"));
+        assertTrue(versionRange.test("1.1.0"));
+        assertTrue(versionRange.test("1.899.0"));
+        assertFalse(versionRange.test("2.0"));
+        assertFalse(versionRange.test("2.0.0"));
+        assertFalse(versionRange.test("2.0.1"));
+    }
+
+    @Test
+    public void testUnboundedRange() {
+        assertThrows(IllegalArgumentException.class, () -> VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "[1.0,2.0"));
+    }
+
+    @Test
+    public void testAlphaVersion() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "[1.0,)");
+        assertFalse(versionRange.test("1.0.0.Alpha"));
+    }
+
+    @Test
+    public void testAlphaVersionInbound() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "[1.0.0.Alpha1,)");
+        assertTrue(versionRange.test("1.0.0.Alpha1"));
+        assertTrue(versionRange.test("1.0.0.Beta"));
+    }
+
+    @Test
+    public void testAlphaVersionInboundExclusive() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "(1.0.0.Alpha1,)");
+        assertFalse(versionRange.test("1.0.0.Alpha1"));
+        assertTrue(versionRange.test("1.0.0.Beta"));
+    }
+
+    @Test
+    public void testMultipleRanges() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "(,1.0],[1.2,)");
+        // Should return true for Versions up to 1.0 (included) and 1.2 or higher
+        assertTrue(versionRange.test("1.0.0.Alpha1"));
+        assertTrue(versionRange.test("1.0.0"));
+        assertFalse(versionRange.test("1.1.0"));
+        assertTrue(versionRange.test("1.2.0"));
+        assertFalse(versionRange.test("1.2.0.Alpha1"));
+    }
+
+    @Test
+    public void testQualifiers() {
+        VersionRange versionRange = VersionRange.createFromVersionSpec(VersionScheme.MAVEN, "[3.8,3.8.5)");
+        assertTrue(versionRange.test("3.8.4.SP1-redhat-00001"));
+        assertTrue(versionRange.test("3.8.4.SP2-redhat-00001"));
+        assertTrue(versionRange.test("3.8.4.redhat-00002"));
+        assertFalse(versionRange.test("3.8.5.redhat-00003"));
```

Comment:
Thanks, this looks good. What if you test `3.8.0.SP1-redhat-00001` for `[3.8.0.redhat-00001,)`?

Reply:
Hm, that fails because the algorithm handles `SP1` as higher than `redhat` (alphabetical order). 

This returns `1`: 

```java
VersionScheme.MAVEN.compare("3.8.0.redhat-00001", "3.8.0.SP1-redhat-00001")
```

Reply:
I wonder if we shouldn't have a specific Red Hat `VersionScheme` for cases like these

Reply:
It's ok, it's not a blocker, I was just wondering. Dependabot has the same issue. We can fix it later. It'd be good to fix it though :)

Reply:
Maven resolver actually does handle it properly.

Reply:
Hm, that sounds like a bug in `MavenVersionScheme`. WDYT @dmlloyd ?

Reply:
I believe that custom qualifiers come after predefined qualifiers (like `SP`) according to the Maven rules.

```
 $ jbang --main=org.eclipse.aether.util.version.GenericVersionScheme org.apache.maven.resolver:maven-resolver-util:1.9.6 1.redhat 1.sp
Display parameters as parsed by Maven Resolver (in canonical form and as a list of tokens) and comparison result:
1. 1.redhat -> 1.redhat; tokens: [1, redhat]
   1.redhat > 1.sp
2. 1.sp -> 1.sp; tokens: [1, 1]
```
```
$ jbang --main=org.eclipse.aether.util.version.GenericVersionScheme org.apache.maven.resolver:maven-resolver-util:1.9.6 3.8.0.redhat-00001 3.8.0.SP1-redhat-00001
Display parameters as parsed by Maven Resolver (in canonical form and as a list of tokens) and comparison result:
1. 3.8.0.redhat-00001 -> 3.8.0.redhat-00001; tokens: [3, 8, redhat, 1]
   3.8.0.redhat-00001 > 3.8.0.SP1-redhat-00001
2. 3.8.0.SP1-redhat-00001 -> 3.8.0.SP1-redhat-00001; tokens: [3, 8, 1, 1, redhat, 1]
```

Reply:
Make sure you aren't using the utilities in `maven-artifact` to test, because that is generally considered to be deprecated at this point (and IIUC is not used by Maven anymore).